GeneralJS/top-product.ts

```diff
@@ -89,6 +89,43 @@ const orders = [
 ];
 
 function topProduct(state: string) {
+  // match state to state code
+  const stateCode = states.find(stateObject => stateObject.state == state)?.stateCode;
+
+  // all products for passed state
+  const stateOrders = orders.filter(order => order.stateCode == stateCode);
+
+  // dict to hold productId : number of products
+  const sortedProducts: any = {};
+
+  // iterate through state orders to add orders to sorted products
+  for(let i = 0; i < stateOrders.length; i++) {
+    let currentProductId = stateOrders[i].productId;
+    // if the product is already present in dictionary add to number of products with that id
+    if(currentProductId in sortedProducts) {
+      sortedProducts[currentProductId] = sortedProducts[currentProductId] + 1;
+    } else {
+      sortedProducts[currentProductId] = 1;
+    }
+  }
+
+  /*
+    Sorted Products
+    {
+      '1': 2, 
+      '2': 1 
+    }
+  */
+
+  // find the product id with the highest number of products
+  // full transparency I cobbled this together from a couple of stack overflow comments also I'm still iterating twice (embarrassing)
+  const topProductId = Number(Object.keys(sortedProducts).reduce((a, b) => sortedProducts[a] > sortedProducts[b] ? a : b));
+  
+  // match product id to product details
+  const topProductDetails = products.find(product => product.productId == topProductId);
```

Comment:
You can also tell me the time complexity of this find

Reply:
O(n). As complex as the array is long but if the matching element is early in the array is can be faster than if it's at the end (assuming the array is a substantial length)